### PR TITLE
Lifetime

### PR DESCRIPTION
**Pull request by andersio**
Built on top of @sharplet's #3030.

Introduced a `LifetimeProviding` protocol which allows objects to declare their ability to notify others of its deinitialization. So far only `NSObject` conforms to this protocol.

`Property` is an ideal participant, but its lifetime semantic & fixes have to be settled first IMO. May as well do that separately.

### CODE REVIEW

**Comment by andersio** [comment]
If we provide only a hot lifetime signal, composed properties are forced to create a hot lifetime signal that *could* be leaked.

So introducing `lifetimeProducer` at least provides a chance for the composed properties to workaround this if they are used with cold signals, assuming we just implement the fix in #3042 without changing the semantic.

File: ReactiveCocoa/Swift/Lifetime.swift [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import enum Result.NoError
+
+/// Describes the ability of an object to notify others of its deinitialization.
+/// Proxy objects may pass through the lifetime of its underlying object.
+public protocol LifetimeProviding: class {
+	/// A signal representing the lifetime of `self`.
+	///
+	/// The signal emits `completed` when the object completes, or
+	/// `interrupted` after the object is completed.
```
Comment (JaviSoto) [inline ReactiveCocoa/Swift/Lifetime.swift]: This sounds a bit confusing

File: ReactiveCocoa/Swift/Signal.swift [hunk 2]
```diff
@@ -791,6 +791,12 @@ extension SignalProtocol {
 			.map { $0.0 }
 	}
 
+	/// Forwards events from `self` until `lifetime` ends, at which point the
+	/// returned signal will complete.
+	public func take<U: LifetimeProviding>(withinLifetime object: U) -> Signal<Value, Error> {
```
Comment (JaviSoto) [inline ReactiveCocoa/Swift/Signal.swift]: Should the parameter be `withinLifetimeOf:`?

**Comment by JaviSoto** [comment]
I like this!

I think we should be able to do it for any `AnyObject`: by holding a `weak` reference to it. Something like:

```swift
final class DeallocationObserver: LifetimeProviding {
    weak var object: AnyObject? {
        didSet {
            if object == nil {
                self.lifetimeObserver.sendCompleted()
            }
        }
    }

    let (lifetime, lifetimeObserver) = Signal<(), NoError>.pipe()

    init(object: AnyObject) {
        self.object = object
    }
}
```

We would just need to make sure the `DeallocationObserver` instance stays alive.

**Comment by andersio** [comment]
Property observer would not be called on weak references turning `nil`.

**Comment by sharplet** [comment]
Yeah, the only way to find out if a weak reference is `nil` is to try to unwrap it.

**Comment by sharplet** [comment]
This is why #3030 is implemented the way it is: so that you can do something like

```swift
class Foo {
  private let lifetime = Lifetime()
}
```

and have something that will be implicitly deallocated at the same time as `Foo`.

**Comment by JaviSoto** [comment]
> Property observer would not be called on weak references turning nil. :-[

Nooooooooooooooooo :(

File: ReactiveCocoa/Swift/Signal.swift (hunk 2 above)
Comment (andersio) [inline ReactiveCocoa/Swift/Signal.swift]: `take(ifRetained:)`? #NamingHardestProblem

Comment (sharplet) [inline ReactiveCocoa/Swift/Signal.swift]: My vote is for `take(withinLifetimeOf:)`.

**Comment by andersio** [comment]
@sharplet We may add a `LifetimeTokenProviding` protocol that accepts a `DeallocationToken` instead of a signal, and let it provide default impl. for `LifetimeProviding`. But it is purely sugar in the sense that it reduces 6 lines of code into one.

**Comment by mdiep** [comment]
#3030 seems much simpler than this and thus preferable since: (a) Swift adoption becomes a single line of code and (b) the `NSObject` implementation is also simpler.

Adding a protocol does seem like it might be worthwhile though.

**Comment by andersio** [comment]
The NSObject implementation is different in only returning the signal or the `Lifetime` object (which is made `internal` in this patch). The objc_sync calls have to be added anyway, since there is no thread safe guarantees on the ObjC association APIs.

**Comment by andersio** [comment]
A protocol like `LifetimeProviding` might be less useful, if `Lifetime` is part of the public API. Say if we write a function `func bind(to: Something, during lifetime: Lifetime`, we are already implicitly constraining to objects that can provide a lifetime.

Having said that, my second thought says `.take(during: myObject.lifetime)` does look better than the super long label suggested here. So I gonna close this.

**Pull request closed without merging**